File: src/simulator_new/link.py
import random
from typing import Optional

from simulator_new.clock import ClockObserver
from simulator_new.trace import Trace

class Link(ClockObserver):
    def __init__(self, id, bw_trace: Optional[Trace] = None,
                 prop_delay_ms=25, queue_cap_bytes=-1,
                 pkt_loss_rate=0) -> None:
        self.id = id
        self.bw_trace = bw_trace
        self.prop_delay_ms = prop_delay_ms
        self.queue_cap_bytes = queue_cap_bytes
        self.queue_size_bytes = 0
        self.pkt_loss_rate = pkt_loss_rate
        self.ts_ms = 0
        self.queue = []
        self.ready_pkts = []
        self.budget_bytes = 0
        self.last_budget_update_ts_ms = 0
        self.host = None

    def register_host(self, host):
        self.host = host

    def push(self, pkt) -> None:
        """Push a packet onto the link"""
        if random.random() < self.pkt_loss_rate:
            return
        if self.queue_cap_bytes == -1 or \
            pkt.size_bytes + self.queue_size_bytes <= self.queue_cap_bytes:
            pkt.add_prop_delay_ms(self.prop_delay_ms)
            if self.bw_trace is None:
                self.ready_pkts.append(pkt)
            else:
                self.queue.append(pkt)
                self.queue_size_bytes += pkt.size_bytes
        else:
            assert self.host
            self.host.cc.on_pkt_lost(self.ts_ms, pkt)
        #     for i in range(len(self.queue)):
        #         print(self.queue[i].pkt_id, self.queue[i].ts_sent_ms, self.queue[i].delay_ms(), self.queue[i].ts_sent_ms + self.queue[i].delay_ms())
        #     print("drop", self.ts_ms, pkt.pkt_id,
        #           pkt.size_bytes + self.queue_size_bytes, self.queue_cap_bytes,
        #           pkt.app_data)

    def pull(self):
        """Pull a packet from the link"""
        # check pkt timestamp to determine whether to dequeue a pkt
        if self.ready_pkts and \
            self.ready_pkts[0].ts_sent_ms + self.ready_pkts[0].delay_ms() <= self.ts_ms:
            return self.ready_pkts.pop(0)
        return None

    def update_bw_budget(self):
        if not isinstance(self.bw_trace, Trace):
            return
        while self.queue:
            pkt = self.queue[0]
            prev_ts_ms = max(pkt.ts_sent_ms, self.last_budget_update_ts_ms)
            delta = int(self.bw_trace.get_avail_bits2send(
                prev_ts_ms / 1000, self.ts_ms / 1000)) / 8
            if prev_ts_ms == pkt.ts_sent_ms:
                self.budget_bytes = delta
            else:
                self.budget_bytes += delta
            self.last_budget_update_ts_ms = self.ts_ms
            if self.budget_bytes >= pkt.size_bytes:
                self.budget_bytes -= pkt.size_bytes
                pkt.add_queue_delay_ms(self.ts_ms - pkt.ts_sent_ms)
                self.queue.pop(0)
                self.queue_size_bytes -= pkt.size_bytes
                self.ready_pkts.append(pkt)
            else:
                break

    def tick(self, ts_ms) -> None:
        assert ts_ms >= self.ts_ms
        if self.ts_ms == ts_ms:
            return
        self.ts_ms = ts_ms
        self.update_bw_budget()

    def reset(self) -> None:
        if isinstance(self.bw_trace, Trace):
            self.bw_trace.reset()
        self.ts_ms = 0
        self.queue = []
        self.queue_size_bytes = 0
        self.budget_bytes = 0
        self.last_budget_update_ts_ms = 0
        self.ready_pkts = []

File: src/simulator/link.cc
#include "link.h"
#include "timestamp.h"
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <limits>
#include <memory>
#include <numeric>
#include <sstream>
#include <string>

Link::Link(const char* fname)
    : qsize_byte_(0), last_budget_update_ts_(0), budget_bit_(0) {
  std::string line;
  unsigned int cnt = 0;
  std::ifstream fin(fname, std::ios::in);
  while (std::getline(fin, line)) {
    if (cnt == 0) {
      cnt++;
      // skip header
      continue;
    }
    std::stringstream ss(line);
    unsigned int col_cnt = 0;
    while (ss.good()) {
      std::string col;
      getline(ss, col, ',');
      switch (col_cnt) {
      case 0:
        ts_.emplace_back(1000 * std::stoul(col));
        break;
      case 1:
        bw_mbps_.emplace_back(std::stod(col));
        break;
      case 2:
        if (!col.empty()) {
          prop_delay_ms_ = std::stoul(col);
        }
        break;
      case 3:
        if (!col.empty()) {
          random_loss_rate_ = std::stod(col);
        }
        break;
      case 4:
        if (!col.empty()) {
          qcap_byte_ = std::stoul(col);
        }
        break;
      case 5:
        // TODO: T_s
        break;
      default:
        break;
      }
      col_cnt++;
    }
    cnt++;
  }
}

double Link::GetAvgBwMbps() const {
  if (bw_mbps_.empty()) {
    return 0.0;
  }
  auto const count = static_cast<double>(bw_mbps_.size());
  return std::reduce(bw_mbps_.cbegin(), bw_mbps_.cend()) / count;
}

void Link::Push(std::unique_ptr<Packet> pkt) {
  const unsigned int pkt_size = pkt->GetSizeByte();
  // drop pkt if random packet loss or queue is full
  // queue capacity is inifite if capacity equals to unsigned int max
  if (static_cast<double>(rand()) / RAND_MAX > random_loss_rate_ &&
      (qcap_byte_ == std::numeric_limits<unsigned int>::max() ||
       pkt_size + qsize_byte_ <= qcap_byte_)) {
    pkt->AddPropDelayMs(prop_delay_ms_);
    queue_.push_back(std::move(pkt));
    qsize_byte_ += pkt_size;
  } else {
    // std::cout << "drop " << pkt->GetSeqNum() << std::endl;
  }
}

std::unique_ptr<Packet> Link::Pull() {
  // Pull a packet from the link
  // check pkt timestamp to determine whether to dequeue a pkt
  if (!ready_pkts_.empty()) {
    std::unique_ptr<Packet>& pkt = ready_pkts_.front();
    const Timestamp& now = Clock::GetClock().Now();
    if (pkt->GetTsSent() +
            TimestampDelta::FromMilliseconds(pkt->GetDelayMs()) <=
        now) {
      auto ret = std::move(ready_pkts_.front());
      ready_pkts_.pop_front();
      return ret;
    }
  }
  return std::unique_ptr<Packet>(nullptr);
}

void Link::Tick() { UpdateBwBudget(); }

void Link::Reset() {
  qsize_byte_ = 0;
  budget_bit_ = 0;
  last_budget_update_ts_.SetUs(0);
  queue_.clear();
  ready_pkts_.clear();
}

unsigned int Link::GetAvailBitsToSend(const Timestamp& t0,
                                      const Timestamp& t1) const {
  const TimestampDelta step = ts_.at(1) - ts_.at(0);
  const Timestamp ts_start = ts_.at(0);
  const int start = ((t0 - ts_start) / step) % ts_.size();
  const int end = ((t1 - ts_start) / step) % ts_.size();
  double bits = 0.0;
  if (start > end) {
    bits =
        std::accumulate(bw_mbps_.cbegin() + start, bw_mbps_.cend(), 0.0) *
            step.ToMicroseconds() +
        std::accumulate(bw_mbps_.cbegin(), bw_mbps_.cbegin() + end + 1, 0.0) *
            step.ToMicroseconds();
  } else {
    bits = std::accumulate(bw_mbps_.cbegin() + start,
                           bw_mbps_.cbegin() + end + 1, 0.0) *
           step.ToMicroseconds();
  }
  bits -= bw_mbps_.at(start) *
          (t0 - (Timestamp::Zero() + step * start)).ToMicroseconds();
  bits -= bw_mbps_.at(end) *
          ((Timestamp::Zero() + step * (end + 1)) - t1).ToMicroseconds();
  // std::cout << "now=" << Clock::GetClock().Now().ToMicroseconds()
  //           << ", ts_start=" << ts_start.ToMicroseconds()
  //           << ", t0=" << t0.ToMicroseconds()
  //           << ", start=" << start //(step * start).ToMicroseconds()
  //           << ", bin start=" << (Timestamp::Zero() + step *
  //           start).ToMicroseconds()
  //           << ", t1=" << t1.ToMicroseconds()
  //           << ", bin end=" << (Timestamp::Zero() + step *
  //           end).ToMicroseconds()
  //           << ", end=" << end // (step * end).ToMicroseconds()
  //           << ", step=" << step.ToMicroseconds()
  //           << ", bits=" << bits
  //           << ", buget_bits=" << budget_bit_
  //           << ", qlen=" << queue_.size()
  //           << std::endl;
  return bits;
}

void Link::UpdateBwBudget() {
  const Timestamp& now = Clock::GetClock().Now();

  while (!queue_.empty()) {
    std::unique_ptr<Packet>& pkt = queue_.front();
    Timestamp ts_sent = pkt->GetTsSent();
    Timestamp prev_ts = std::max(ts_sent, last_budget_update_ts_);
    const unsigned int pkt_size = pkt->GetSizeByte();
    const auto bits = GetAvailBitsToSend(prev_ts, now);
    budget_bit_ = prev_ts == ts_sent ? bits : budget_bit_ + bits;
    last_budget_update_ts_ = now;
    if (budget_bit_ >= (pkt_size * 8)) {
      budget_bit_ -= (pkt_size * 8);
      pkt->AddQueueDelayMs((now - ts_sent).ToMilliseconds());
      qsize_byte_ -= pkt_size;
      ready_pkts_.push_back(std::move(pkt));
      queue_.pop_front();
    } else {
      break;
    }
  }
}
